// @jorgebodega/typeorm-seeding/dist/commands/config.command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigCommand = void 0;
const yargs_1 = require("yargs");
const chalk_1 = require("chalk");
const connection_1 = require("../connection");
class ConfigCommand {
    constructor() {
        this.command = 'config';
        this.describe = 'Show the TypeORM config with seeding values';
    }
    /**
     * @inheritdoc
     */
    builder(args) {
        return args
            .option('n', {
            alias: 'configName',
            type: 'string',
            describe: 'Name of the typeorm config file (json or js).',
        })
            .option('c', {
            alias: 'connection',
            type: 'string',
            default: 'default',
            describe: 'Name of the typeorm connection',
        })
            .option('r', {
            alias: 'root',
            type: 'string',
            describe: 'Path to your typeorm config file',
        });
    }
    /**
     * @inheritdoc
     */
    async handler(args) {
        try {
            (0, connection_1.configureConnection)({
                root: args.root,
                configName: args.configName,
                connection: args.connection,
            });
            const options = await (0, connection_1.getConnectionOptions)();
            console.log(options);
        }
        catch (error) {
            console.log('\n❌ ', (0, chalk_1.red)('Could not find the orm config file'));
            console.error(error);
            (0, yargs_1.exit)(1, error);
        }
    }
}
exports.ConfigCommand = ConfigCommand;
